core/domain/story_services.py

```diff
@@ -559,6 +559,34 @@ def get_completed_node_ids(user_id, story_id):
     return progress_model.completed_node_ids if progress_model else []
 
 
+def get_latest_completed_node_ids(user_id, story_id):
+    """Returns the ids of the completed nodes that are latest in the story.
+
+    Args:
+        user_id: str. ID of the given user.
+        story_id: str. ID of the story.
+
+    Returns:
+        list(str). List of the completed node ids latest in story. If length
+        is larger than 3, return the lastest three of them. If length is smaller
+        or equal to 3, return all of them.
+    """
+    progress_model = user_models.StoryProgressModel.get(
+        user_id, story_id, strict=False)
+
+    if not progress_model:
+        return []
+
+    num_of_nodes = min(len(progress_model.completed_node_ids), 3)
+    latest_completed_node_ids = []
+    for node_id in progress_model.completed_node_ids:
```

Comment:
As mentioned before, do not assume anything about the ordering or composition of completed_node_ids. Get the numbers for each node, sort them, then take the nodes corresponding to the largest three.

What you have currently doesn't work for cases like [7, 2, 3, 4], I think.

Reply:
Done.